File: mobile/lib/models/vehicle.dart
class Vehicle {
  final String id;
  final String make;
  final String model;
  final String? variant;
  final int year;
  final String licensePlate;
  final String? type;
  final String? status;
  final String? image;
  final VehicleInsurance? insurance;
  final List<VehicleDocument> documents;
  final VehicleLocation? location;
  final DateTime? lastService;
  final DateTime? nextService;
  final String? vin;
  final num? mileage;
  final String? fuelType;
  final String? color;
  final String? frontTyres;
  final String? rearTyres;
  final String? batteryDetails;
  final Map<String, dynamic>? healthIndicators;

  Vehicle({
    required this.id,
    required this.make,
    required this.model,
    this.variant,
    required this.year,
    required this.licensePlate,
    this.type,
    this.status,
    this.image,
    this.insurance,
    this.documents = const [],
    this.location,
    this.lastService,
    this.nextService,
    this.vin,
    this.mileage,
    this.fuelType,
    this.color,
    this.frontTyres,
    this.rearTyres,
    this.batteryDetails,
    this.healthIndicators,
  });

  factory Vehicle.fromJson(Map<String, dynamic> json) {
    return Vehicle(
      id: (json['id'] ?? json['_id'] ?? '').toString(),
      make: (json['make'] ?? '').toString(),
      model: (json['model'] ?? '').toString(),
      variant: json['variant']?.toString(),
      year: (json['year'] is num) ? (json['year'] as num).toInt() : 0,
      licensePlate: (json['licensePlate'] ?? '').toString(),
      type: json['type']?.toString(),
      status: json['status']?.toString(),
      image: json['image']?.toString(),
      vin: json['vin']?.toString(),
      mileage: json['mileage'] is num ? (json['mileage'] as num) : null,
      fuelType: json['fuelType']?.toString(),
      color: json['color']?.toString(),
      frontTyres: json['frontTyres']?.toString(),
      rearTyres: json['rearTyres']?.toString(),
      batteryDetails: json['batteryDetails']?.toString(),
      healthIndicators: json['healthIndicators'] is Map
          ? Map<String, dynamic>.from(json['healthIndicators'])
          : null,
      lastService: json['lastService'] != null
          ? DateTime.tryParse(json['lastService'].toString())
          : null,
      nextService: json['nextService'] != null
          ? DateTime.tryParse(json['nextService'].toString())
          : null,
      insurance: json['insurance'] != null
          ? VehicleInsurance.fromJson(
              Map<String, dynamic>.from(json['insurance']),
            )
          : null,
      documents: (json['documents'] as List? ?? [])
          .map((e) => VehicleDocument.fromJson(Map<String, dynamic>.from(e)))
          .toList(),
      location: json['location'] != null
          ? VehicleLocation.fromJson(
              Map<String, dynamic>.from(json['location']),
            )
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'make': make,
      'model': model,
      if (variant != null) 'variant': variant,
      'year': year,
      'licensePlate': licensePlate,
      if (type != null) 'type': type,
      if (status != null) 'status': status,
      if (image != null) 'image': image,
      if (vin != null) 'vin': vin,
      if (mileage != null) 'mileage': mileage,
      if (fuelType != null) 'fuelType': fuelType,
      if (color != null) 'color': color,
      if (frontTyres != null) 'frontTyres': frontTyres,
      if (rearTyres != null) 'rearTyres': rearTyres,
      if (batteryDetails != null) 'batteryDetails': batteryDetails,
      if (lastService != null) 'lastService': lastService?.toIso8601String(),
      if (nextService != null) 'nextService': nextService?.toIso8601String(),
      if (insurance != null) 'insurance': insurance!.toJson(),
      'documents': documents.map((e) => e.toJson()).toList(),
      if (location != null) 'location': location!.toJson(),
    };
  }
}

class VehicleInsurance {
  final String? policyNumber;
  final String? provider;
  final DateTime? startDate;
  final DateTime? expiryDate;
  final String? status;

  VehicleInsurance({
    this.policyNumber,
    this.provider,
    this.startDate,
    this.expiryDate,
    this.status,
  });

  factory VehicleInsurance.fromJson(Map<String, dynamic> json) {
    return VehicleInsurance(
      policyNumber: json['policyNumber']?.toString(),
      provider: json['provider']?.toString(),
      startDate: json['startDate'] != null
          ? DateTime.tryParse(json['startDate'].toString())
          : null,
      expiryDate: json['expiryDate'] != null
          ? DateTime.tryParse(json['expiryDate'].toString())
          : null,
      status: (json['status'] ?? 'Active').toString(),
    );
  }

  Map<String, dynamic> toJson() => {
    'policyNumber': policyNumber,
    'provider': provider,
    'startDate': startDate?.toIso8601String(),
    'expiryDate': expiryDate?.toIso8601String(),
    'status': status,
  };
}

class VehicleDocument {
  final String name;
  final String type;
  final String url;
  final DateTime? expiryDate;
  final DateTime? uploadedAt;

  VehicleDocument({
    required this.name,
    required this.type,
    required this.url,
    this.expiryDate,
    this.uploadedAt,
  });

  factory VehicleDocument.fromJson(Map<String, dynamic> json) {
    return VehicleDocument(
      name: (json['name'] ?? '').toString(),
      type: (json['type'] ?? 'Other').toString(),
      url: (json['url'] ?? '').toString(),
      expiryDate: json['expiryDate'] != null
          ? DateTime.tryParse(json['expiryDate'].toString())
          : null,
      uploadedAt: json['uploadedAt'] != null
          ? DateTime.tryParse(json['uploadedAt'].toString())
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
    'name': name,
    'type': type,
    'url': url,
    'expiryDate': expiryDate?.toIso8601String(),
    'uploadedAt': uploadedAt?.toIso8601String(),
  };
}

class VehicleLocation {
  final String? address;
  final double? lat;
  final double? lng;

  VehicleLocation({this.address, this.lat, this.lng});

  factory VehicleLocation.fromJson(Map<String, dynamic> json) {
    return VehicleLocation(
      address: json['address']?.toString(),
      lat: (json['lat'] as num?)?.toDouble(),
      lng: (json['lng'] as num?)?.toDouble(),
    );
  }

  Map<String, dynamic> toJson() => {'address': address, 'lat': lat, 'lng': lng};
}

File: backend/models/Vehicle.js
import mongoose from 'mongoose';

const vehicleSchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    required: true,
    ref: 'User',
  },
  make: { type: String, required: true },
  model: { type: String, required: true },
  variant: { type: String },
  year: { type: Number, required: true },
  licensePlate: { type: String, required: true },
  registrationDate: { type: String },
  color: { type: String },
  image: { type: String }, // URL
  vin: { type: String },
  mileage: { type: Number },
  fuelType: { type: String },
  frontTyres: { type: String },
  rearTyres: { type: String },
  batteryDetails: { type: String },
  type: { type: String, enum: ['Car'], default: 'Car' },
  status: { type: String, enum: ['Idle', 'On Route', 'In Service'], default: 'Idle' },
  location: {
    lat: { type: Number },
    lng: { type: Number },
    address: { type: String },
    updatedAt: { type: Date }
  },
  geo: {
    type: {
      type: String,
      default: 'Point',
      enum: ['Point']
    },
    coordinates: {
      type: [Number],
      index: '2dsphere'
    }
  },
  lastService: { type: Date },
  nextService: { type: Date },
  documents: [{
    name: { type: String },
    type: { type: String, enum: ['Insurance', 'Registration', 'Battery Warranty', 'Tire Warranty', 'Other'] },
    url: { type: String },
    expiryDate: { type: Date },
    uploadedAt: { type: Date, default: Date.now }
  }],
  insurance: {
    policyNumber: { type: String },
    provider: { type: String },
    startDate: { type: Date },
    expiryDate: { type: Date },
    status: { type: String, enum: ['Active', 'Expired', 'Expiring Soon'], default: 'Active' },
    claims: [{
      claimNumber: { type: String },
      date: { type: Date, default: Date.now },
      description: { type: String },
      amount: { type: Number },
      status: { type: String, enum: ['Pending', 'Approved', 'Rejected'], default: 'Pending' }
    }]
  },
  healthIndicators: {
    generalService: {
      value: { type: Number, default: 0 },
      lastUpdated: { type: Date, default: Date.now },
      lastServiceDate: { type: Date },
      lastServiceKm: { type: Number, default: 0 },
      fixedKm: { type: Number, default: 0 },
      fixedDays: { type: Number, default: 0 }
    },
    brakePads: {
      value: { type: Number, default: 0 },
      lastUpdated: { type: Date, default: Date.now },
      lastServiceDate: { type: Date },
      lastServiceKm: { type: Number, default: 0 },
      fixedKm: { type: Number, default: 0 },
      fixedDays: { type: Number, default: 0 }
    },
    tires: {
      value: { type: Number, default: 0 },
      lastUpdated: { type: Date, default: Date.now },
      lastServiceDate: { type: Date },
      lastServiceKm: { type: Number, default: 0 },
      fixedKm: { type: Number, default: 0 },
      fixedDays: { type: Number, default: 0 }
    },
    battery: {
      value: { type: Number, default: 0 },
      lastUpdated: { type: Date, default: Date.now },
      lastServiceDate: { type: Date },
      lastServiceKm: { type: Number, default: 0 },
      fixedKm: { type: Number, default: 0 },
      fixedDays: { type: Number, default: 0 }
    },
    wiperBlade: {
      value: { type: Number, default: 0 },
      lastUpdated: { type: Date, default: Date.now },
      lastServiceDate: { type: Date },
      lastServiceKm: { type: Number, default: 0 },
      fixedKm: { type: Number, default: 0 },
      fixedDays: { type: Number, default: 0 }
    }
  }
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Helper to calculate health
vehicleSchema.methods.calculateHealth = function() {
  const vehicle = this;
  if (!vehicle.healthIndicators) return;

  const now = new Date();
  const currentKm = vehicle.mileage || 0;

  const keys = ['generalService', 'brakePads', 'tires', 'battery', 'wiperBlade'];
  
  keys.forEach(key => {
    const indicator = vehicle.healthIndicators[key];
    if (indicator && indicator.lastServiceDate) {
      const lastDate = new Date(indicator.lastServiceDate);
      const diffTime = Math.abs(now - lastDate);
      // We use floor here because Day 1 should be 0 days passed
      const diffDays = Math.floor(diffTime / (1000 * 60 * 60 * 24));
      
      const lastKm = indicator.lastServiceKm || 0;
      const diffKm = Math.max(0, currentKm - lastKm);
      const fixedKm = indicator.fixedKm || 0;
      const fixedDays = indicator.fixedDays || 0;

      // Calculate service progress based on days (if fixedDays > 0)
      const progressFromDays = fixedDays > 0 ? Math.min(100, (diffDays / fixedDays) * 100) : 0;
      
      // Calculate service progress based on KM (if fixedKm > 0)
      const progressFromKm = fixedKm > 0 ? Math.min(100, (diffKm / fixedKm) * 100) : 0;

      // Service progress is the maximum of both (whichever comes first)
      indicator.value = Math.round(Math.max(progressFromDays, progressFromKm));
    } else if (indicator) {
      // If no service date has been set by a merchant, all indicators must be zero
      indicator.value = 0;
      indicator.fixedKm = 0;
      indicator.fixedDays = 0;
    }
  });
};

vehicleSchema.pre('save', async function() {
  this.calculateHealth();
});

vehicleSchema.index({ user: 1, createdAt: -1 });

const Vehicle = mongoose.model('Vehicle', vehicleSchema);

export default Vehicle;
